src/core/data/ndarray/ndarray_vector.h

```diff
@@ -5,275 +5,128 @@
-        }
 
-        NdArrayVector2D()                              = delete;
-        NdArrayVector2D(NdArrayVector2D const& source) = default;
-        NdArrayVector2D(NdArrayVector2D&& source)      = default;
-        NdArrayVector2D& operator                      =(NdArrayVector2D const& source)
+        NdArrayVector& operator=(NdArrayVector const& source)
         {
-            if (nx_ != source.nx_ || ny_ != source.ny_)
+            if (nCells_ != source.nCells_)
             {
                 throw std::runtime_error(
-                    "Error NdArrayVector1D cannot be assigned, incompatible sizes");
-            }
-            else
-            {
-                this->data_ = source.data_;
+                    "Error NdArrayVector cannot be assigned, incompatible sizes");
             }
+
+            this->data_ = source.data_;
             return *this;
         }
 
-        NdArrayVector2D& operator=(NdArrayVector2D&& source)
+        NdArrayVector& operator=(NdArrayVector&& source)
         {
-            if (nx_ != source.nx_ || ny_ != source.ny_)
+            if (nCells_ != source.nCells_)
             {
                 throw std::runtime_error(
-                    "Error NdArrayVector1D cannot be assigned, incompatible sizes");
-            }
-            else
-            {
-                this->data_ = std::move(source.data_);
+                    "Error NdArrayVector cannot be assigned, incompatible sizes");
             }
-            return *this;
-        }
-
-        //! read/write data access operator returns C-ordered data.
-        DataType& operator()(uint32_t i, uint32_t j) { return this->data_[j + ny_ * i]; }
 
-        //! read only access. See read/write.
-        DataType const& operator()(uint32_t i, uint32_t j) const
-        {
-            return this->data_[linearIt(i, j)];
+            this->data_ = std::move(source.data_);
+            return *this;
         }
 
-        static const std::size_t dimension = 2;
-        using type                         = DataType;
-
-    private:
-        int constexpr linearIt(int i, int j) const { return j + ny_ * i; }
-
-
-        uint32_t nx_ = 0;
-        uint32_t ny_ = 0;
-    };
-
 
-
-    //! NdArrayVector3D is an implementation for a 3-dimensional container
-    /** behaves as the 2D version.
-     */
-    template<typename DataType = double>
-    class NdArrayVector3D : public NdArrayVectorBase<DataType>
-    {
-    public:
-        NdArrayVector3D(uint32_t nx, uint32_t ny, uint32_t nz)
-            : NdArrayVectorBase<DataType>(nx * ny * nz)
-            , nx_{nx}
-            , ny_{ny}
-            , nz_{nz}
+        template<typename... Indexes>
+        DataType const& operator()(Indexes... indexes) const
         {
-        }
+            auto params = std::tuple<Indexes...>{indexes...};
+            static_assert(sizeof...(Indexes) == dim);
+            // static_assert((... && std::is_unsigned_v<decltype(indexes)>)); TODO : manage later if
```

Comment:
perhaps a >=0 check could suffice?

Reply:
I agree. I tried something like
`static_assert((... && std::greater_equal<int32_t>()), 0);`
which does not compile, but I can hardly figure out the way to write it...

Reply:
I don't think it can be a compile time check, as it's a runtime value check.

```
 static_assert((... && !std::is_floating_point_v<Indexes>));
 assert ((... && (indexes >= 0)));
```